greeter/README.md

```diff
@@ -1,21 +1,428 @@
+
+    ```ts
+    import {inject} from '@loopback/context';
+    import {CachingService} from '../services/caching.service';
+    import {CACHING_SERVICE} from '../keys';
+    ```
+
+3. For the `start()` and `stop()` function, we want to start and stop the caching service.
+
+    ```ts
+    /**
+     * This method will be invoked when the application starts
+     */
+    async start(): Promise<void> {
+        // Add your logic for start
+        await this.cachingService.start();
+    }
+
+    /**
+     * This method will be invoked when the application stops
+     */
+    async stop(): Promise<void> {
+        // Add your logic for stop
+        await this.cachingService.stop();
+    }
+    ```
+
+
+### Step 3: Create global interceptor for caching
+
+1. Run the interceptor generator `lb4 interceptor` command.
+
+    ```sh
+    $ lb4 interceptor
+    ? Interceptor name: caching
+    ? Is it a global interceptor? Yes
+
+    Global interceptors are sorted by the order of an array of group names bound to ContextBindings.GLOBAL_INTERCEPTOR_ORDERED_GROUPS. See https://loopback.io/doc/en/lb4/Interceptors.html#order-of-invocation-for-interceptors.
+
+    ? Group name for the global interceptor: ('') 
+    create src/interceptors/caching.interceptor.ts
+    update src/interceptors/index.ts
+
+    Interceptor Caching was created in src/interceptors/
+    ```
+
+2. In `src/interceptors/caching.interceptors.ts`, modify the constructor:
+
+    ```ts
+    constructor(
+        @inject(CACHING_SERVICE) private cachingService: CachingService,
+    ) {}
+    ```
+
+    Add and modify the following import statements:
+
+    ```ts
+    import {
+    /* inject, */
+    globalInterceptor,
+    Interceptor,
+    InvocationContext,
+    InvocationResult,
+    Provider,
+    inject,
+    ValueOrPromise,
+    } from '@loopback/context';
+    import {CachingService} from '../services/caching.service';
+    import {CACHING_SERVICE} from '../keys';
+    import {RestBindings} from '@loopback/rest';
+    ```
+
+3. Add the pre-invocation and post-invocation logics in the `intercept` function.
+
+    ```ts
+    // Add pre-invocation logic here
+    const httpReq = await invocationCtx.get(RestBindings.Http.REQUEST, {
+        optional: true,
+    });
+    /* istanbul ignore if */
+    if (!httpReq) {
+        // Not http request
+        return next();
+    }
+    const key = httpReq.path;
+    const lang = httpReq.acceptsLanguages(['en', 'zh']) || 'en';
+    const cachingKey = `${lang}:${key}`;
+    const cachedResult = await this.cachingService.get(cachingKey);
+    if (cachedResult) {
+        console.error('Cache found for %s %j', cachingKey, cachedResult);
+        return cachedResult;
+    }
+
+    const result = await next();
+
+    // Add post-invocation logic here
+    await this.cachingService.set(cachingKey, result);
+    return result;
+    ```
+
+### Step 4: Try it out again! 
+
+Restart your application. 
+
+_TODO: how to test caching is working?_
```

Comment:
@jannyHou, probably need your help here.  How do we know the caching is enabled?

Reply:
### Test cache

The expire time for each result is set to 10s (10000 milliseconds by default).
To test the cached result and its expiration, you can 

1. Try `GET/ greet/{name}` with a name 'loopback' and language 'en'
2. You will see a result object with a timestamp field like
```
{
  "timestamp": "2019-10-09T20:24:00.294Z",
  "language": "en",
  "greeting": "Hello, loopback!"
}
```
3. Try the same endpoint and same input immediately (before the cached result expires), you will see the same timestamp printed in the result, which infers that the result is NOT calculated again but from cache.
4. Wait for 10s and try same endpoint again, you will see the timestamp changes, like
```
{
  // please note the time here changes
  "timestamp": "2019-10-09T20:30:29.770Z",
  "language": "en",
  "greeting": "Hello, loopback!"
}
```

Summary: this is the functionality enabled by adding the cache interceptor.

### Test sweep

The sweep interval is set to 10s(10,000 milliseconds) too.
Since the cache sweep runs every 10s in the background, we can observe the sweep function gets executed by turning on the debug string and restart the app.

To test the cache sweeping, you can:

1. Stop your app and run it again with debug string turning on:
```
// exit your app
ctrl + c
// start it again with debug
DEBUG=greeter-extension npm run start
```

2. Watch the console window to see how the sweeping function is triggered every 10s
```
greeter-extension Sweeping cache... +10s
greeter-extension Sweeping cache... +10s
greeter-extension Sweeping cache... +10s
greeter-extension Sweeping cache... +10s
greeter-extension Sweeping cache... +10s
```
3. Try endpoint `GET /greet/{name}` again from explorer, see its result
4. Go back to the console window, after 10s, you will see the cached result gets swept
```
Cache for en:/greet/loopback is swept.
```

Reply:
I also see 

```
greeter-extension Caching options: undefined +0ms
```